### ISSUE TITLE
Textbox is giving very slow response

### ISSUE THREAD

**Issue by ali-gshare** [julmot/datatables.mark.js#1]
when searching the textbox gives very slow response in Chrome. Whole textbox freezes while typing. Is it possible to make it fast? http://codepen.io/anon/pen/GqWQBJ

**Comment by julmot** [comment]
Hi @ali-gshare,

Thanks for submitting!

Please provide the following information:

1. What kind of mark.js version you are using (as stated in the [contribution guideline](https://github.com/julmot/datatables.mark.js/blob/master/CONTRIBUTING.md#11-questions-problems-and-bugs))
2. Do you really need to show 500 entries per page?

Could you please also clean up your codepen? Please put the JS and CSS things into the related sections, remove `<meta>`, `<html>`, `<body>` and `<head>` tags and include external dependencies in the behavior tab:
![2016-06-28_235245](https://cloud.githubusercontent.com/assets/8310677/16433617/7595ce02-3d8b-11e6-9f27-9f1c7a8be0f8.png)

Thank you in advance.

**Comment by julmot** [comment]
@ali-gshare As you don't reply I've taken this myself and reproduced it in a clean way in this fiddle: 
https://jsfiddle.net/julmot/u7xgzr61/

It's not a Chrome specific issue, I can also reproduce it in Firefox. However, when disabling datatables.mark.js the search is already a bit slow. So, this issue is due to the huge amount of rows you've chosen – 500. When setting it to e.g. 50 you'll not have any issues.

Anyway, to not worsen the already slow performance of DataTables in this situation, the plugin could check the current entries in the instance and if it's over a specific point, not listen to each keypress but in a defined interval.

What do you think?

**Comment by ali-gshare** [comment]
Hi @julmot Thanks for your reply. Unfortunately client requirement is 500 rows. Yes listening to interval will help. Do you have any idea where i could start looking to hack this or you could share some sample code with me. Many thanks!

**Comment by DataTables** [comment]
The pen linked to above is nearly instant for me on my Mac Mini. The first character when searching is a little slow - but that's not surprising given them about of DOM manipulation that Mark.js is doing to highlight a single character in so many different locations. Disabling Mark.js and its searching as fast as I can type.

If you used paging the Mark.js could update the DOM on only the visible elements which would make things massively faster.

**Comment by julmot** [comment]
@ali-gshare Thank you for your answer.

> Do you have any idea where i could start looking to hack this or you could share some sample code with me. Many thanks!

There's nothing necessary to "hack". I'd just like to discuss possible solutions and then implement the best one directly into the plugin to make it generally available. If you're using Bower you'll automatically receive this update.

@DataTables Thank you for participation.

> The pen linked to above is nearly instant for me on my Mac Mini. 

I'm surprised about this. I've tested it on two different laptops:

- Win 10, Intel i5, 16 GB RAM
- Win 7, Intel i7, 8 GB RAM

and the fiddle I've posted is slow in both of them – regardless if mark.js is enabled or not, but it's even slower when enabled.

> If you used paging the Mark.js could update the DOM on only the visible elements which would make things massively faster.

Currently, it checks for the elements on the current page only. I personally don't think that just highlighting the elements that are visible in the viewport and then observing scroll events to continue highlighting will be an improvement. From a usability point of view this would mean that highlights will be "lazy" – especially on iPad, where the scroll event is only fired when completed – and scroll smoothness would be poor on touch devices like Android tablets. Highlighting matches should be a visual help for users to faster find what they were searching for. When they need to wait for highlights or if the browser freezes on slow devices the target would not be achieved.

Checking in an interval at a specific amount of elements per page is the only way I currently can imagine to solve this issue. But I'll have to test this first on my browser/device farm in the next week (including a Mac Mini).

Or are there any other ideas?

**Comment by julmot** [comment]
@ali-gshare @DataTables Good news. I've just released [v1.0.1](https://github.com/julmot/datatables.mark.js/releases/tag/v1.0.1) fixing this issue – within the bounds of possibility.

Please have a look at [this fiddle](https://jsfiddle.net/u7xgzr61/1/) with the new release. As you can see the plugin detects that there are 50+ rows per page and will only highlight matches when the user stopped entering. Previously it highlighted matches on each input change. 
This causes the highlighting process to only run one time when entering something. The remaining slowness is due to DataTables itself. As you can see in [this fiddle](https://jsfiddle.net/u7xgzr61/2/), it is also slow when datatables.mark.js is deactivated. 

I've tested on a few devices and came to the following result:

- On iPad 3 with iOS 8.4, having 500 rows per page will crash the browser (Safari) entirely
- On iPad 3 the DataTables search also with e.g. 50 rows per page is very slow
- The new datatables.mark.js release may detect that the user stopped entering something on iPad 3 even if this isn't the case, as the browser often freezes due to the fact that the DataTables filter action in the background is slow and the interval is executed before DataTables has finished filtering
- On a Nexus tablet running Android 5 having 500 rows per page will be very slow, but will not crash the browser. The slowness even exists when datatables.mark.js is deactivated.
- On a Surface 3 running Windows 10 having 500 rows per page will cause the browser (Edge) to hide the keyboard after entering one character, even if datatables.mark.js is deactivated. When having 100 rows per page the search is still slow, but the keyboard will not disappear.
- I can confirm that the performance is better on Mac Mini than on Windows laptops. However, the new release works great on both.

In summary: This release solves the slowness of datatables.mark.js on large pages (> 49), but can't improve the performance of DataTables itself. Therefore having 500 rows may still be slow, but not due to datatables.mark.js. On desktop browsers you probably can live with that because it's not dramatic, but when targeting mobile devices you should be careful with the number of rows per page.

Btw: I've tested with keywords "alloy" and "resistant".

**Comment by ali-gshare** [comment]
@julmot Thanks very much for your amazing work. I will test and update you if any feedback. Many thanks!

**Comment by julmot** [comment]
@ali-gshare Thank you for your answer.  Please let me know if this solves the issue in your application.
